Java/examples/src/main/java/com/amazon/randomcutforest/examples/dynamicinference/ConditionalPredictive.java

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amazon.randomcutforest.examples.dynamicinference;
+
+import static com.amazon.randomcutforest.CommonUtils.checkArgument;
+import static java.lang.Math.min;
+
+import java.io.BufferedWriter;
+import java.io.FileWriter;
+import java.util.Random;
+
+import com.amazon.randomcutforest.PredictiveRandomCutForest;
+import com.amazon.randomcutforest.config.TransformMethod;
+import com.amazon.randomcutforest.examples.Example;
+import com.amazon.randomcutforest.returntypes.SampleSummary;
+import com.amazon.randomcutforest.summarization.Summarizer;
+
+public class ConditionalPredictive implements Example {
+
+    public static void main(String[] args) throws Exception {
+        new ConditionalPredictive().run();
+    }
+
+    @Override
+    public String command() {
+        return "Conditional_predictive_example";
+    }
+
+    @Override
+    public String description() {
+        return "An example that uses imputation for prediction";
+    }
+
+    @Override
+    public void run() throws Exception {
+        // Create and populate a random cut forest
+
+        int shingleSize = 1;
+        int numberOfTrees = 100;
+        int sampleSize = 256;
+        int dataSize = 40 * sampleSize;
+
+        // 5 dimensions, three are known and 4,5 th unknown (and stochastic)
+        int baseDimensions = 5;
+
+        PredictiveRandomCutForest forest = new PredictiveRandomCutForest.Builder<>().inputDimensions(baseDimensions)
+                .randomSeed(0).numberOfTrees(numberOfTrees).shingleSize(shingleSize).sampleSize(sampleSize)
+                .startNormalization(sampleSize / 2).transformMethod(TransformMethod.NORMALIZE).build();
+
+        long seed = 17;
+        new Random().nextLong();
+
+        System.out.println("seed = " + seed);
+        NormalDistribution normal = new NormalDistribution(new Random(seed));
+        Random random = new Random(seed + 10);
+
+        String name = "predictive_example";
+        BufferedWriter file = new BufferedWriter(new FileWriter(name));
+        for (int i = 0; i < dataSize; i++) {
+            float[] record = generateRecordKey(random);
+            checkArgument(record[3] == 0, " should not be filled");
+            checkArgument(record[4] == 0, " should not be filled");
+
+            SampleSummary answer = forest.predict(record, 0, new int[] { 3, 4 });
```

Comment:
Can we think predict multiple missing values multivariate forecasting? You told me multivariate forecasting is complicated and rcf hasn't supported it. If yes, how can we predict multiple missing values confidently?

I am thinking it aloud for the potential integration on AD plugin as follows.  

In the context of streaming supervised learning, this can be used for multi-output regression (e.g., When the outputs or targets are continuous values. For example, predicting the temperature and humidity levels for the next day based on today's weather data.) or multi-label classification (e.g., When each instance can belong to multiple classes. For instance, a text document might be about both 'science' and 'technology', which are not mutually exclusive categories.), right?

We can use a multi-target regression or classification rcf model depending on the nature of the KPIs. For example, a multi-output regression model might predict both the latency and error rate simultaneously. The model's input features would include the change indicators. When the model detects a pattern that historically correlates with certain types of system degradations, it can make predictions about the current state. When increased latency and error rates are observed, rcf can evaluate the likelihood of these being related to recent changes such as a specific deployment or a configuration change and we can find root cause using rcf's attribution.

Of course, rcf can be used to output not only the predicted KPIs but also the likelihood of various potential root causes. This can be done by structuring the output layer to have nodes corresponding to various root causes, trained to activate when the input data matches historical patterns leading to those causes. For instance, rcf might indicate that there is an 80% likelihood that the current latency issue is due to a recent deployment and a 50% likelihood that it's related to a configuration change.

Reply:
> Can we think predict multiple missing values multivariate forecasting? You told me multivariate forecasting is  > complicated and rcf hasn't supported it. If yes, how can we predict multiple missing values confidently?

Actually RCF never disallowed it :) But I would always recommend using it with caution. Multivariate forecasting is slightly different from multi target regression. The issue is that in simple algorithms for Multivariate forecasting the multiple targets are then used to generate subsequent predictions -- and that can be a self reinforcing loop. 
For one dimension this can mostly be controlled -- but with multiple dimensions it becomes near infeasible.  More complicated algorithms also have the same problem in varying degrees (c.f. hallucinations). But multi-target regression in itself is fine -- in fact RCF can perform tree regression (just like RFs) and streaming comes for free. So a forecast horizon of 1 is not a big issue at all. 

> I am thinking it aloud for the potential integration on AD plugin as follows.

We should try it out.